flytekit/remote/remote.py

```diff
@@ -700,10 +700,17 @@ def _serialize_and_register(
                 )
 
             try:
+                c_version = version
+                if isinstance(cp_entity, task_models.TaskSpec):
+                    if cp_entity.template and cp_entity.template.container and cp_entity.template.container.image:
+                        # Add the image version to the workflow hash
+                        # TODO: does settings.version need to be updated?
+                        tmp_hash = cp_entity.template.container.image.split(":")[-1].split(".")[0]
+                        c_version = f"{version}{tmp_hash}"
                 ident = self.raw_register(
                     cp_entity,
                     settings=settings,
-                    version=version,
+                    version=c_version,
                     create_default_launchplan=create_default_launchplan,
```

Comment:
Could you please explain why this is required?
My naive expectation after reading the PR description is that the plugin "should do the`docker build -t <some tag> .` part" that has to come before `pyflyte run --image <some tag>`. Doing this manually wouldn't require modifying the version here, right?

Reply:
The version hash is built based on the python code - which doesn't include the Dockerfile if it changes.  This means that if you `pyflyte run` the same workflow twice but the container hash changes due to a change in the dockerfile, you errors similar to this:
```
RPC Failed, with Status: StatusCode.INVALID_ARGUMENT
        details: task with different structure already exists with id resource_type:TASK project:"flytesnacks" domain:"development" name:"example_pyflyte_script.t1" version:"LMxQheYT9myAmWklNHYD6A=="
        ```

Reply:
yeah, I think workflow and task version will become different

Reply:
hash is calculated [here](https://github.com/flyteorg/flytekit/blob/085bc645749de66eb1deae5ccd40284b38b821e9/flytekit/remote/remote.py#L779), we can add the dockerfile to the zip file. Hence, flytekit will register a new version of workflow once you update the dockerfile.

Reply:
a problem is that we would have to also hash the docker build context directory as well

Reply:
Or we could do it [here](https://github.com/flyteorg/flytekit/blob/38c76876dfe7fc2c62536ca6a195bce8a56c6270/flytekit/remote/remote.py#L837)?

```py
@staticmethod
    def _version_from_hash(
        md5_bytes: bytes,
        serialization_settings: SerializationSettings,
        *additional_context: str,
    ) -> str:
    """
    ...
    :param additional_context: This is for additional context to factor into the version computation,
          meant for objects (like Options for instance) that don't easily consistently stringify.
```

`pyflyte run` calls [`register_script`](https://github.com/flyteorg/flytekit/blob/38c76876dfe7fc2c62536ca6a195bce8a56c6270/flytekit/remote/remote.py#L920) [under the hood](https://github.com/flyteorg/flytekit/blob/38c76876dfe7fc2c62536ca6a195bce8a56c6270/flytekit/clis/sdk_in_container/run.py#L490), which does:

```py
if version is None:
    # The md5 version that we send to S3/GCS has to match the file contents exactly,
    # but we don't have to use it when registering with the Flyte backend.
    # For that add the hash of the compilation settings to hash of file
    version = self._version_from_hash(md5_bytes, serialization_settings)
```

If the user on the other hand [executes `pyflyte register`](https://github.com/flyteorg/flytekit/blob/38c76876dfe7fc2c62536ca6a195bce8a56c6270/flytekit/clis/sdk_in_container/register.py#L174), [under the hood we do](https://github.com/flyteorg/flytekit/blob/38c76876dfe7fc2c62536ca6a195bce8a56c6270/flytekit/tools/repo.py#L246):

```py
 if not version and fast:
        version = remote._version_from_hash(md5_bytes, serialization_settings, service_account, raw_data_prefix)  # noqa
        click.secho(f"Computed version is {version}", fg="yellow")
```

Reply:
The tag isn't generated/parsed until the end of register_script (which uses the previously defined version to create some entities)  I guess you could do it in 2 stages & with a junk version initially, and then change the version on the resulting entities?

<details>
  <summary>build image backtrace</summary>
 
  
  ```
> /Users/dan.farrell/tmp/git/flytekit/flytekit/image_spec/image_spec.py(160)build()                                                                                                                      
-> if image_spec.builder not in cls._REGISTRY:                                                                                                                                                           
(Pdb) bt                                                                                                                                                                                                 
  /Users/dan.farrell/.local/share/micromamba/envs/tflyte-subwf-01/bin/pyflyte(8)<module>()                                                                                                               
-> sys.exit(main())                                                                                                                                                                                      
  /Users/dan.farrell/.local/share/micromamba/envs/tflyte-subwf-01/lib/python3.11/site-packages/click/core.py(1157)__call__()                                                                             
-> return self.main(*args, **kwargs)                                                                                                                                                                     
  /Users/dan.farrell/.local/share/micromamba/envs/tflyte-subwf-01/lib/python3.11/site-packages/rich_click/rich_group.py(21)main()                                                                        
-> rv = super().main(*args, standalone_mode=False, **kwargs)                                                                                                                                             
  /Users/dan.farrell/.local/share/micromamba/envs/tflyte-subwf-01/lib/python3.11/site-packages/click/core.py(1078)main()                                                                                 
-> rv = self.invoke(ctx)                                                                                                                                                                                 
  /Users/dan.farrell/tmp/git/flytekit/flytekit/clis/sdk_in_container/utils.py(124)invoke()                                                                                                               
-> return super().invoke(ctx)                                                                                                                                                                            
  /Users/dan.farrell/.local/share/micromamba/envs/tflyte-subwf-01/lib/python3.11/site-packages/click/core.py(1688)invoke()                                                                               
-> return _process_result(sub_ctx.command.invoke(sub_ctx))                                                                                                                                               
  /Users/dan.farrell/.local/share/micromamba/envs/tflyte-subwf-01/lib/python3.11/site-packages/click/core.py(1688)invoke()                                                                               
-> return _process_result(sub_ctx.command.invoke(sub_ctx))                                                                                                                                               
  /Users/dan.farrell/.local/share/micromamba/envs/tflyte-subwf-01/lib/python3.11/site-packages/click/core.py(1688)invoke()                                                                               
-> return _process_result(sub_ctx.command.invoke(sub_ctx))                                                                                                                                               
  /Users/dan.farrell/.local/share/micromamba/envs/tflyte-subwf-01/lib/python3.11/site-packages/click/core.py(1434)invoke()                                                                               
-> return ctx.invoke(self.callback, **ctx.params)                                                                                                                                                        
  /Users/dan.farrell/.local/share/micromamba/envs/tflyte-subwf-01/lib/python3.11/site-packages/click/core.py(783)invoke()                                                                                
-> return __callback(*args, **kwargs)                                                                                                                                                                    
  /Users/dan.farrell/tmp/git/flytekit/flytekit/clis/sdk_in_container/run.py(490)_run()                                                                                                                   
-> remote_entity = remote.register_script(                                                                                                                                                               
  /Users/dan.farrell/tmp/git/flytekit/flytekit/remote/remote.py(932)register_script()                                                                                                                    
-> return self.register_task(entity, serialization_settings, version)                                                                                                                                    
  /Users/dan.farrell/tmp/git/flytekit/flytekit/remote/remote.py(735)register_task()                                                                                                                      
-> ident = self._serialize_and_register(entity=entity, settings=serialization_settings, version=version)                                                                                                 
  /Users/dan.farrell/tmp/git/flytekit/flytekit/remote/remote.py(691)_serialize_and_register()                                                                                                            
-> _ = get_serializable(m, settings=serialization_settings, entity=entity, options=options)                                                                                                              
  /Users/dan.farrell/tmp/git/flytekit/flytekit/tools/translator.py(696)get_serializable()                                                                                                                
-> cp_entity = get_serializable_task(settings, entity)                                                                                                                                                   
  /Users/dan.farrell/tmp/git/flytekit/flytekit/tools/translator.py(180)get_serializable_task()                                                                                                           
-> container = entity.get_container(settings)                                                                                                                                                            
  /Users/dan.farrell/tmp/git/flytekit/flytekit/core/python_auto_container.py(177)get_container()                                                                                                         
-> return self._get_container(settings)                                                                                                                                                                  
  /Users/dan.farrell/tmp/git/flytekit/flytekit/core/python_auto_container.py(188)_get_container()                                                                                                        
-> image=get_registerable_container_image(self.container_image, settings.image_config),                                                                                                                  
  /Users/dan.farrell/tmp/git/flytekit/flytekit/core/python_auto_container.py(257)get_registerable_container_image()
-> ImageBuildEngine.build(img)
> /Users/dan.farrell/tmp/git/flytekit/flytekit/image_spec/image_spec.py(160)build()
-> if image_spec.builder not in cls._REGISTRY:
(Pdb)
  ```
  
</details>